python/lsst/pipe/tasks/processCcdWithFakes.py

```diff
@@ -343,29 +359,38 @@ def runQuantum(self, butlerQC, inputRefs, outputRefs):
             expId, expBits = butlerQC.quantum.dataId.pack("visit_detector", returnMaxBits=True)
             inputs['exposureIdInfo'] = ExposureIdInfo(expId, expBits)
 
+        expWcs = inputs["exposure"].getWcs()
+        tractId = inputs["skyMap"].findTract(
+            expWcs.pixelToSky(inputs["exposure"].getBBox().getCenter())).tract_id
         if not self.config.doApplyExternalGlobalSkyWcs and not self.config.doApplyExternalTractSkyWcs:
-            inputs["wcs"] = inputs["exposure"].getWcs()
-
+            inputs["wcs"] = expWcs
         elif self.config.doApplyExternalGlobalSkyWcs:
             externalSkyWcsCatalog = inputs["externalSkyWcsGlobalCatalog"]
             row = externalSkyWcsCatalog.find(detectorId)
             inputs["wcs"] = row.getWcs()
-
         elif self.config.doApplyExternalTractSkyWcs:
-            externalSkyWcsCatalog = inputs["externalSkyWcsTractCatalog"]
+            externalSkyWcsCatalogList = inputs["externalSkyWcsTractCatalog"]
+            for externalSkyWcsCatalogRef in externalSkyWcsCatalogList:
+                if externalSkyWcsCatalogRef.dataId["tract"] == tractId:
+                    externalSkyWcsCatalog = externalSkyWcsCatalogRef.get(
+                        datasetType=self.config.connections.externalSkyWcsTractCatalog)
+                    break
```

Comment:
@morriscb I have seen this return in some places with no `externalSkyWcsCatalog` because the condition is never satisfied. You could let it use the last `externalSkyWcsCatalog` with an else statement like:

```
else:
    externalSkyWcsCatalog = externalSkyWcsCatalogRef.get(datasetType=self.config.connections.externalSkyWcsTractCatalog)
```

Similar problem for the photoCalib below in addition to a bug (see below L391).

Reply:
+1 for @surhudm's thinking above - it would be good for this large if/elif block to have a fall-back option in place.

Reply:
If the condition is never satisfied, that means there is no wcs or photocalib for that tract. I'm not sure what the safest fall back position. Likely people doing who have written the analysis code and use these calibs should decide.